src/solvers/strings/string_format_builtin_function.cpp

```diff
@@ -0,0 +1,482 @@
+/*******************************************************************\
+
+Module: Builtin functions for String.format
+
+Author: Romain Brenguier, Joel Allred
+
+Date:   June 2019
+
+\*******************************************************************/
+
+/// \file
+///  Builtin functions for String.format
+
+#include <math.h>
+#include <string>
+
+#include "format_specifier.h"
+#include "string_format_builtin_function.h"
+#include <util/range.h>
+#include <util/simplify_expr.h>
+#include <util/symbol_table.h>
+
+string_format_builtin_functiont::string_format_builtin_functiont(
+  const exprt &return_code,
+  const std::vector<exprt> &fun_args,
+  array_poolt &array_pool)
+  : string_builtin_functiont(return_code, array_pool)
+{
+  PRECONDITION(fun_args.size() >= 3);
+  result = array_pool.find(fun_args[1], fun_args[0]);
+  const array_string_exprt format_string_expr =
+    get_string_expr(array_pool, fun_args[2]);
+
+  // List of arguments after the format string
+  inputs = make_range(fun_args.begin() + 3, fun_args.end())
+             .map([&](const exprt &arg) {
+               INVARIANT(
+                 is_refined_string_type(arg.type()),
+                 "arguments of format should be strings");
+               return get_string_expr(array_pool, arg);
+             })
+             .collect<std::vector<array_string_exprt>>();
+
+  // format_string is only initialized if the expression is constant
+  if(
+    array_pool.get_or_create_length(format_string_expr).id() == ID_constant &&
+    format_string_expr.content().id() == ID_array)
+  {
+    const auto length = numeric_cast_v<std::size_t>(
+      to_constant_expr(array_pool.get_or_create_length(format_string_expr)));
+    format_string = utf16_constant_array_to_java(
+      to_array_expr(format_string_expr.content()), length);
+  }
+}
+
+static std::vector<mp_integer> deserialize_constant_int_arg(
+  const std::vector<mp_integer> serialized,
+  const unsigned base)
+{
+  PRECONDITION(serialized.size() == 4);
+
+  // long value, to be used for other formats?
+  for(int i = 0; i < 5; i++)
+  {
+    DATA_INVARIANT(
+      serialized[i] <= 0xFFFF,
+      "Component of serialized value to"
+      "format must be bounded by 0xFFFF");
+  }
+
+  const int64_t int64_value =
+    (serialized[0] << 48).to_long() | (serialized[1] << 32).to_long() |
+    (serialized[2] << 16).to_long() | serialized[3].to_long();
+  const mp_integer mp_integer_value{int64_value};
+  const std::string long_as_string = integer2string(mp_integer_value, base);
+
+  const std::vector<mp_integer> deserialized =
+    make_range(long_as_string).map([&](char c) { return mp_integer{c}; });
+
+  return deserialized;
```

Comment:
:pick: could return directly the expression above

Reply:
done